src/game.cpp

```diff
@@ -319,176 +321,223 @@ static inline u8 getNeighbors(v3s16 p, INodeDefManager *nodedef, ClientMap *map,
 	return neighbors;
 }
 
-/*
-	Find what the player is pointing at
-*/
-PointedThing getPointedThing(Client *client, Hud *hud, const v3f &player_position,
-		const v3f &camera_direction, const v3f &camera_position,
-		core::line3d<f32> shootline, f32 d, bool liquids_pointable,
+/*!
+ * Returns the object or node the player is pointing at.
+ * Also handles the rendering of the selection.
+ * @param[in]  client            the client object
+ * @param[in]  hud               the game's hud
+ * @param[in]  shootline         the shootline, starting from
+ * the camera position. This also gives the maximal distance
+ * of the search.
+ * @param[in]  liquids_pointable if false, liquids are ignored
+ * @param[in]  look_for_object   if false, objects are ignored
+ * @param[in]  camera_offset     offset of the camera
+ * @param[out] selected_object   the selected object or
+ * NULL if not found
+ */
+PointedThing getPointedThing(Client *client, Hud *hud,
+		core::line3d<f32> shootline, bool liquids_pointable,
 		bool look_for_object, const v3s16 &camera_offset,
 		ClientActiveObject *&selected_object)
 {
 	PointedThing result;
 
 	std::vector<aabb3f> *selectionboxes = hud->getSelectionBoxes();
 	selectionboxes->clear();
-	static const bool show_entity_selectionbox = g_settings->getBool("show_entity_selectionbox");
+	static const bool show_entity_selectionbox = g_settings->getBool(
+		"show_entity_selectionbox");
 
 	selected_object = NULL;
 
 	INodeDefManager *nodedef = client->getNodeDefManager();
+
+	core::aabbox3d<s16> maximal_exceed = nodedef->getSelectionBoxIntUnion();
+	// The code needs to search these nodes
+	core::aabbox3d<s16> search_range(-maximal_exceed.MaxEdge,
+		-maximal_exceed.MinEdge);
+	// If a node is found, there might be a larger node behind.
+	// To find it, we have to go further.
+	s16 maximal_overcheck =
+		  std::max(abs(search_range.MinEdge.X), abs(search_range.MaxEdge.X))
+		+ std::max(abs(search_range.MinEdge.Y), abs(search_range.MaxEdge.Y))
+		+ std::max(abs(search_range.MinEdge.Z), abs(search_range.MaxEdge.Z));
+
 	ClientMap &map = client->getEnv().getClientMap();
 
-	f32 min_distance = BS * 1001;
+	const v3f original_vector = shootline.getVector();
+	const f32 original_length = original_vector.getLength();
+
+	f32 min_distance = original_length;
 
 	// First try to find a pointed at active object
 	if (look_for_object) {
-		selected_object = client->getSelectedActiveObject(d * BS,
-				  camera_position, shootline);
+		selected_object = client->getSelectedActiveObject(shootline,
+			result.intersection_point, result.intersection_normal);
 
 		if (selected_object != NULL) {
-			if (show_entity_selectionbox &&
-					selected_object->doShowSelectionBox()) {
+			if (show_entity_selectionbox
+				&& selected_object->doShowSelectionBox()) {
```

Comment:
the operator should always be at the end, not the start. Plus, it should be indented twice. So the code before was correct, the code after is wrong.